File: src/main/java/ev/projects/Floor.java
package ev.projects;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Floor {
    private final List<Tile> tiles;
    private int firstPlayerMarkerPosition;
    private final Lid lid;

    public Floor() {
        this(new Lid());
    }

    public Floor(Lid lid) {
        tiles = new ArrayList<>();
        firstPlayerMarkerPosition = -1;
        this.lid = lid;
    }

    void clear() {
        tiles.forEach(lid::addTile);
        tiles.clear();
        firstPlayerMarkerPosition = -1;
    }

    void addFirstPlayerMarker() {
        firstPlayerMarkerPosition = tiles.size();
    }

    void add(Tile tile, int amount) {
        int overfill = tiles.size() + amount - 7;
        for (int i = 0; i < overfill; i++) {
            lid.addTile(tile);
        }
        int amountToFill = Math.min(amount, 7 - tiles.size());
        for (int i = 0; i < amountToFill; i++) {
            tiles.add(tile);
        }
    }

    int score() {
        int tileCountWithMarker = (firstPlayerMarkerPosition > -1) ? (tiles.size() + 1) : tiles.size();
        return switch (tileCountWithMarker) {
            case 0 -> 0;
            case 1 -> -1;
            case 2 -> -2;
            case 3 -> -4;
            case 4 -> -6;
            case 5 -> -8;
            case 6 -> -11;
            case 7 -> -14;
            default -> throw new RuntimeException("Unexpected tile count");
        };
    }

    @Override
    public String toString() {
        if (tiles.size() == 0) {
            if (firstPlayerMarkerPosition == -1) {
                return "Empty";
            }
            return "M";
        }
        StringBuilder result = new StringBuilder(
                tiles.stream().map(Tile::toString).collect(Collectors.joining(""))
        );
        if (firstPlayerMarkerPosition > -1) {
            result.insert(firstPlayerMarkerPosition, "M");
        }
        return String.join(" ", result.toString().split(""));
    }

    public List<String> jsonList() {
        List<String> jsonList = new ArrayList<>();
        for (Tile tile : tiles) {
            jsonList.add(tile.toString());
        }
        if (firstPlayerMarkerPosition > -1) {
            jsonList.add(firstPlayerMarkerPosition, "M");
        }
        return jsonList;
    }
}

File: python/src/floor.py
from src.lid import Lid

class Floor:
    def __init__(self, lid=None):
        if lid is None:
            lid = Lid()
        self.tiles = []
        self.first_player_marker_position = -1
        self.lid = lid

    def clear(self):
        for tile in self.tiles:
            self.lid.add_tile(tile)
        self.tiles.clear()
        self.first_player_marker_position = -1

    def add_first_player_marker(self):
        self.first_player_marker_position = len(self.tiles)

    def add(self, tile, amount):
        overfill = len(self.tiles) + amount - 7
        for _ in range(overfill):
            self.lid.add_tile(tile)
        amount_to_fill = min(amount, 7 - len(self.tiles))
        for _ in range(amount_to_fill):
            self.tiles.append(tile)

    def score(self):
        tile_count_with_marker = len(self.tiles) + 1 if self.first_player_marker_position > -1 else len(self.tiles)
        return {
            0: 0,
            1: -1,
            2: -2,
            3: -4,
            4: -6,
            5: -8,
            6: -11,
            7: -14,
        }.get(tile_count_with_marker, RuntimeError("Unexpected tile count"))

    def __str__(self):
        if len(self.tiles) == 0:
            return "Empty" if self.first_player_marker_position == -1 else "M"
        result = "".join(str(tile) for tile in self.tiles)
        if self.first_player_marker_position > -1:
            result = result[:self.first_player_marker_position] + "M" + result[self.first_player_marker_position:]
        return " ".join(result)

    def json_list(self):
        json_list = [str(tile) for tile in self.tiles]
        if self.first_player_marker_position > -1:
            json_list.insert(self.first_player_marker_position, "M")
        return json_list